backfill_rpc/src/rpc.rs

```diff
@@ -8,26 +17,80 @@ use interface::{
     error::UsecaseError, slot_getter::FinalizedSlotGetter, solana_rpc::TransactionsGetter,
 };
 use plerkle_serialization::serializer::seralize_encoded_transaction_with_status;
+#[cfg(test)]
+use solana_client::rpc_sender::RpcSender;
 use solana_client::{
-    nonblocking::rpc_client::RpcClient, rpc_client::GetConfirmedSignaturesForAddress2Config,
+    client_error::{ClientError, ClientErrorKind},
+    nonblocking::rpc_client::RpcClient,
+    rpc_client::GetConfirmedSignaturesForAddress2Config,
     rpc_config::RpcTransactionConfig,
+    rpc_request::RpcError,
 };
 use solana_program::pubkey::Pubkey;
 use solana_sdk::{
     commitment_config::{CommitmentConfig, CommitmentLevel},
     signature::Signature,
 };
 use solana_transaction_status::UiTransactionEncoding;
+use tokio::sync::Notify;
+use tokio_retry::{strategy::ExponentialBackoff, RetryIf};
 
 const MAX_SIGNATURES_LIMIT: usize = 50_000_000;
-pub(crate) const GET_TX_RETRIES: usize = 7;
+const INITIAL_BACKOFF_DELAY_MS: u64 = 1_000;
+pub(crate) const MAX_RPC_RETRIES: usize = 7;
 pub struct BackfillRPC {
-    pub(crate) client: Arc<RpcClient>,
+    pub(crate) client: BackoffRpcClient,
 }
 
 impl BackfillRPC {
     pub fn connect(addr: String) -> Self {
-        Self { client: Arc::new(RpcClient::new(addr)) }
+        let client = RpcClient::new(addr);
+        Self { client: BackoffRpcClient::new_with_default_backoff(client) }
+    }
+
+    pub(crate) async fn get_signatures_by_address(
+        &self,
+        until: Option<Signature>,
+        before: Option<Signature>,
+        address: &Pubkey,
+    ) -> Result<Vec<SignatureWithSlot>, UsecaseError> {
+        let signatures = self
+            .client
+            .execute_with_backoff(|| async {
+                self.client
+                    .get_signatures_for_address_with_config(
+                        address,
+                        GetConfirmedSignaturesForAddress2Config {
+                            until,
+                            commitment: Some(CommitmentConfig {
+                                commitment: CommitmentLevel::Finalized,
+                            }),
+                            before,
+                            ..Default::default()
+                        },
+                    )
+                    .await
+            })
+            .await;
+        signatures
+            .map_err(Into::<UsecaseError>::into)?
+            .into_iter()
+            .map(|response| {
+                Ok(SignatureWithSlot {
+                    signature: Signature::from_str(&response.signature)?,
+                    slot: response.slot,
+                })
+            })
+            .collect::<Result<Vec<_>, UsecaseError>>()
+    }
+
+    #[cfg(test)]
+    pub(crate) fn new_with_sender_and_constant_backoff(
```

Comment:
Can't we just inline the function in the place where it is needed? At first glance, placing it in a trait and hiding it behind a feature flag doesn't look appropriate.

Reply:
even though it was in a struct, i agree